streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobGenerator.scala

```diff
@@ -43,6 +47,9 @@ class JobGenerator(jobScheduler: JobScheduler) extends Logging {
   private val ssc = jobScheduler.ssc
   private val conf = ssc.conf
   private val graph = ssc.graph
+  private lazy val completedBothCheckpoints = new HashMap[Time, Boolean]()
```

Comment:
I dont like the fact that the data structure encodes that there will be two checkpoints for each batch. Checkpointing is something that should be doable any time, and any number of times. Restricting a data strcuture to a constraint that it can only be done twice is bad design.

In fact I say that my earlier design of automatically deleting checkpoints whenever checkpoint is called, was wrong as well. That's why adding the flag in my PR is important, so that we dissociate the act of taking a checkpoint from the act of clearing checkpoints.